File: web/src/models/Calendar.ts
// Conforms to docs/data/calendar.md, spec v1.
// The lunar-calendar overlay: maps a civil date to its Gaudiya lunar month, and that month
// to songs. Holds no titles - `songUids` resolve through the Manifest, exactly as SongGroup
// does (docs/data/collections.md).
import { Uid } from './Common'

/**
 * Why a song is attached to a month. Lets UI rank or filter by evidence strength rather
 * than treating a well-attested seasonal song and a thematic guess as equivalent.
 */
export type CalendarBasis = 'observed' | 'panjika' | 'book' | 'thematic'

/** A song reference plus the evidence backing its placement. */
export interface ICalendarSongRef {
  uid: Uid
  basis: CalendarBasis
}

/**
 * One pūrṇimānta lunar month as a HALF-OPEN date span: `start` inclusive, `end` exclusive,
 * where `end` is always the next window's `start`. Guarantees a date matches exactly one
 * window. Boundary days carry ±1 day uncertainty - see docs/data/calendar.md.
 */
export interface ILunarWindow {
  /** ISO date, inclusive. */
  start: string
  /** ISO date, exclusive. */
  end: string
  /** New moon falling inside this window. */
  newMoon: string
  /** Kārtika, Śrāvaṇa, … */
  lunarMonth: string
  /** Viṣṇu-name: Dāmodara, Śrīdhara, … */
  gaudiyaMonth: string
  /** True for Puruṣottama, the intercalary leap month (~every 2.7 years). */
  adhika: boolean
}

/** Songs and observances for one lunar month. `songs` may legitimately be empty. */
export interface ICalendarMonth {
  lunarMonth: string
  gaudiyaMonth: string
  /** Festivals falling in this month. */
  observances: string[]
  note?: string
  songs: ICalendarSongRef[]
}

/** A time-of-day ārati slot - applies every day, independent of the lunar month. */
export interface ICalendarDailySlot {
  slot: string
  label: string
  songs: ICalendarSongRef[]
}

/**
 * A dated observance (spec v2). Unlike ILunarWindow, which resolves a *month*, these are the
 * actual observance days from the Śrī Caitanya-pañjikā.
 */
export interface ICalendarFestival {
  id: string
  name: string
  /** Gregorian months it can fall in (lunar dates drift, so this may span two). */
  months: string[]
  /** Observance dates, ascending. Always non-empty. */
  dates: string[]
  /** May be empty — many observances have no matching song in the corpus. */
  songs: ICalendarSongRef[]
}

/** A festival paired with its next occurrence on/after a given date. */
export interface IUpcomingFestival {
  festival: ICalendarFestival
  /** ISO date of the next occurrence. */
  date: string
  /** Whole days from the reference date; 0 = today. */
  daysAway: number
}

/** The whole overlay, as bundled in calendar.json. */
export interface ICalendar {
  specVersion: number
  generated: string
  windowStart: string
  windowEnd: string
  windows: ILunarWindow[]
  months: ICalendarMonth[]
  daily: ICalendarDailySlot[]
  festivals: ICalendarFestival[]
  basisLegend: Record<string, string>
}

/** What a date resolves to: which month it falls in, and what to sing. */
export interface ICalendarToday {
  date: string
  window: ILunarWindow
  month: ICalendarMonth
  daily: ICalendarDailySlot[]
}

File: andorid/app/src/main/java/com/gaudiyakirtan/myapplication/models/Calendar.kt
package com.gaudiyakirtan.myapplication.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * The lunar-calendar overlay (docs/data/calendar.md, spec v2), decoded from the bundled
 * `calendar.json`.
 *
 * It is an **overlay, not a catalog entity**: it holds no titles or lyrics, only `song_uids` that
 * resolve through the Manifest — exactly how [SongGroup] references songs. Field names mirror
 * web's `models/Calendar.ts` so the same concept reads the same way on every platform.
 *
 * The lunar month is an astronomical quantity; the pipeline precomputes the windows so the three
 * platforms do a date-range lookup instead of three drifting floating-point ports.
 */
@Serializable
data class GaudiyaCalendar(
    @SerialName("spec_version") val specVersion: Int = 0,
    @SerialName("generated") val generated: String = "",
    @SerialName("window_start") val windowStart: String = "",
    @SerialName("window_end") val windowEnd: String = "",
    @SerialName("windows") val windows: List<LunarWindow> = emptyList(),
    @SerialName("months") val months: List<CalendarMonth> = emptyList(),
    @SerialName("daily") val daily: List<CalendarDailySlot> = emptyList(),
    @SerialName("festivals") val festivals: List<CalendarFestival> = emptyList()
)

/**
 * One pūrṇimānta lunar month as a **half-open** span: [start] inclusive, [end] exclusive, where
 * [end] equals the next window's [start]. That guarantees a date matches exactly one window.
 */
@Serializable
data class LunarWindow(
    @SerialName("start") val start: String,
    @SerialName("end") val end: String,
    @SerialName("new_moon") val newMoon: String = "",
    @SerialName("lunar_month") val lunarMonth: String,
    @SerialName("gaudiya_month") val gaudiyaMonth: String,
    /** True for Puruṣottama, the intercalary leap month (~every 2.7 years). */
    @SerialName("adhika") val adhika: Boolean = false
)

/** Songs and observances for one lunar month. [songs] may legitimately be empty (e.g. Pauṣa). */
@Serializable
data class CalendarMonth(
    @SerialName("lunar_month") val lunarMonth: String,
    @SerialName("gaudiya_month") val gaudiyaMonth: String,
    @SerialName("observances") val observances: List<String> = emptyList(),
    @SerialName("note") val note: String? = null,
    @SerialName("songs") val songs: List<CalendarSongRef> = emptyList()
)

/**
 * A song reference plus the evidence backing its placement, so the UI can rank by evidence
 * strength rather than treating a well-attested seasonal song and a thematic guess as equal.
 */
@Serializable
data class CalendarSongRef(
    @SerialName("uid") val uid: String,
    @SerialName("basis") val basis: String = "thematic"
)

/** A time-of-day ārati slot — applies every day, independent of the lunar month. */
@Serializable
data class CalendarDailySlot(
    @SerialName("slot") val slot: String,
    @SerialName("label") val label: String = "",
    @SerialName("songs") val songs: List<CalendarSongRef> = emptyList()
)

/** A dated observance from the Śrī Caitanya-pañjikā. */
@Serializable
data class CalendarFestival(
    @SerialName("id") val id: String,
    @SerialName("name") val name: String,
    @SerialName("months") val months: List<String> = emptyList(),
    @SerialName("dates") val dates: List<String> = emptyList(),
    @SerialName("songs") val songs: List<CalendarSongRef> = emptyList()
)

/** What to sing on a given date: the window it falls in and that month's songs/observances. */
data class CalendarToday(
    val date: String,
    val window: LunarWindow,
    val month: CalendarMonth
)

/**
 * Evidence-strength ranking (docs/screens/today.md "Ranking and provenance"): strongest first,
 * `observed` → `panjika` → `book` → `thematic`.
 *
 * An `observed` song is backed by dated recordings of it actually being sung that month; a
 * `thematic` one is a subject-matter match with no dated evidence. Ranking by this is required;
 * *displaying* the basis is not.
 */
fun basisRank(basis: String): Int = when (basis) {
    "observed" -> 0
    "panjika" -> 1
    "book" -> 2
    else -> 3
}
